shaders/src/getters.vs

```diff
@@ -62,18 +63,84 @@ vec3 mulBoneVertex(vec3 v, uint i) {
     return v.x * m[0].xyz + (v.y * m[1].xyz + (v.z * m[2].xyz + m[3].xyz));
 }
 
+void skinPosition(inout vec3 p, const uvec4 ids, const vec4 weights) {
+    // standard skinning for 4 weights, some of them could be zero
+    if (weights.w >= 0.0) {
+        p = weights.x * mulBoneVertex(p, uint(ids.x))
+            + weights.y * mulBoneVertex(p, uint(ids.y))
+            + weights.z * mulBoneVertex(p, uint(ids.z))
+            + weights.w * mulBoneVertex(p, uint(ids.w));
+        return;
+    }
+    // skinning for >4 weights
+    vec3 posSum = weights.x * mulBoneVertex(p, uint(ids.x));
+    posSum += weights.y * mulBoneVertex(p, uint(ids.y));
+    posSum += weights.z * mulBoneVertex(p, uint(ids.z));
+    uint pairIndex = uint(-weights.w - 1.);
+    uint pairStop = pairIndex + uint(ids.w - 3u);
+    for (uint i = pairIndex; i < pairStop; i = i + 1u) {
+        ivec2 texcoord = ivec2(i % MAX_SKINNING_BUFFER_WIDTH, i / MAX_SKINNING_BUFFER_WIDTH);
+        vec2 indexWeight = texelFetch(bonesBuffer_indicesAndWeights, texcoord, 0).rg;
+        posSum += mulBoneVertex(p, uint(indexWeight.r)) * indexWeight.g;
+    }
+    p = posSum;
+}
+
 void skinNormal(inout vec3 n, const uvec4 ids, const vec4 weights) {
-    n =   mulBoneNormal(n, ids.x) * weights.x
-        + mulBoneNormal(n, ids.y) * weights.y
-        + mulBoneNormal(n, ids.z) * weights.z
-        + mulBoneNormal(n, ids.w) * weights.w;
+    // standard skinning for 4 weights, some of them could be zero
+    if (weights.w >= 0.0) {
+        n = weights.x * mulBoneNormal(n, uint(ids.x))
+            + weights.y * mulBoneNormal(n, uint(ids.y))
+            + weights.z * mulBoneNormal(n, uint(ids.z))
+            + weights.w * mulBoneNormal(n, uint(ids.w));
+        return;
+    }
+    // skinning for >4 weights
+    vec3 normSum = weights.x * mulBoneNormal(n, uint(ids.x));
+    normSum += weights.y * mulBoneNormal(n, uint(ids.y));
+    normSum += weights.z * mulBoneNormal(n, uint(ids.z));
+    uint pairIndex = uint(-weights.w - 1.);
+    uint pairStop = pairIndex + uint(ids.w - 3u);
+    for (uint i = pairIndex; i < pairStop; i = i + 1u) {
```

Comment:
any reason to prefer `i = i + 1u` over `++i` ?

Reply:
You are right, it's better, thank you